core/src/main/scala/shapeless/package.scala

```diff
@@ -59,6 +59,60 @@ package object shapeless {
   type ∃[P[_]] = P[T] forSome { type T }
   type ∀[P[_]] = ¬[∃[({ type λ[X] = ¬[P[X]]})#λ]]
 
+  //Type logic-gates
+  trait ||[A, B] extends Serializable {
+    def a: Option[A]
+    def b: Option[B]
+  }
+
```

Comment:
Interesting ... I think I would be tempted to define a `Pad` type class and have separate instances for the three cases ... you should be able to eliminate the pattern match on the options.